File: test/org/cerion/symcalc/expression/function/integer/RandomIntegerTest.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.Environment;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.junit.Test;

import static org.junit.Assert.*;

public class RandomIntegerTest {

    @Test
    public void binary() {
        Expr e = new RandomInteger();

        int[] count = { 0, 0 };

        for(int i = 0; i < 10000; i++) {
            IntegerNum n = (IntegerNum) e.eval();
            int x = n.toInteger();
            count[x]++;
        }

        //if this fails occasionally it might still be okay
        assertTrue("0 count = " + count[0], count[0] > 4900);
        assertTrue("1 count = " + count[1], count[1] > 4900);
    }

    @Test
    public void numbersAreRandom() {
        Expr e = new RandomInteger(new IntegerNum(10));

        IntegerNum prev = new IntegerNum(-99);
        for(int i = 0; i < 5; i++) {
            IntegerNum curr = (IntegerNum)e.eval();
            if(!prev.equals(curr))
                return;

        }

        // All numbers were the same
        assertFalse(true);
    }
}

File: src/test/java/function/statistics/RandomIntegerTest.kt
package org.cerion.symcalc.function.statistics

import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.number.Integer
import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class RandomIntegerTest {

    @Test
    fun binary() {
        val e = RandomInteger()

        val count = intArrayOf(0, 0)

        for (i in 0..9999) {
            val n = e.eval() as Integer
            val x = n.intValue()
            count[x]++
        }

        assertTrue(count[0] > 4800, "0 count = " + count[0])
        assertTrue(count[1] > 4800, "1 count = " + count[1])
    }

    @Test
    fun maxValue() {
        val e = RandomInteger(Integer(10))
        val found = intArrayOf(0, 5, 10)

        for (i in 0..9999) {
            val n = e.eval().asInteger()
            val rand = n.intValue()
            if (rand == 0)
                found[0] = -1
            else if (rand == 5)
                found[1] = -1
            else if (rand == 10)
                found[2] = -1

            if (found[0] == -1 && found[1] == -1 && found[2] == -1)
                return
        }

        assertFalse(true)
    }

    @Test
    fun range() {
        val e = RandomInteger(ListExpr(Integer(10), Integer(20)))
        val found = intArrayOf(10, 15, 20)

        for (i in 0..9999) {
            val n = e.eval().asInteger()
            val rand = n.intValue()
            if (rand == 10)
                found[0] = -1
            else if (rand == 15)
                found[1] = -1
            else if (rand == 20)
                found[2] = -1
            else if (rand < 10)
                assertFalse(true)
            else if (rand > 20)
                assertFalse(true)

            if (found[0] == -1 && found[1] == -1 && found[2] == -1)
                return
        }

        assertFalse(true)
    }

    @Test
    fun numbersAreRandom() {
        val e = RandomInteger(Integer(10))

        val prev = Integer(-99)
        for (i in 0..4) {
            val curr = e.eval() as Integer
            if (!prev.equals(curr))
                return
        }

        // All numbers were the same
        assertFalse(true)
    }
}